src/event-store.ts

```diff
@@ -0,0 +1,234 @@
+      '_eventstore.EventStoreReplayRequested',
+      createEventReplayHandler({store: this.store, emitter: this.emitter}),
+    );
+  }
+
+  public async save(event: Event<EventData, EventContext<any>>): Promise<void> {
+    await this.store.write(event).then((result) => {
+      return result
+        .map(() => {
+          // If there are no errors saving, emit the event
+          return this.emitter.emit(event);
+        })
+        .getOrElseL(() => {
+          return result
+            .swap()
+            .map((error) => {
+              if (error instanceof DuplicateError) {
+                return Promise.resolve();
+              }
+              return Promise.reject(error);
+            })
+            .get();
+        });
+    });
+  }
+
+  public createAggregate<A extends any[], T>(
+    aggregateName: string,
+    query: Q,
+    matches: AggregateMatches<T>,
+  ): Aggregate<A, T> {
+    const _impl = async (...args: A): Promise<Option<T>> => {
+      const start = Date.now();
+
+      const id = createHash('sha256')
+        .update(aggregateName + JSON.stringify(query) + JSON.stringify(args))
+        .digest('hex');
+
+      const latestSnapshot = await this.cache.get<T>(id);
+
+      this.logger.trace('cacheSnapshot', latestSnapshot);
+      const results = this.store.read(
+        query,
+        latestSnapshot.flatMap((snapshot) => Option.of(snapshot.time)),
+        ...args,
+      );
+
+      const aggregatedAt = new Date();
+      const aggregator = composeAggregator(matches);
+
+      const aggregatedResult = await reduce<Event<EventData, EventContext<any>>, Option<T>>(
+        results,
+        latestSnapshot.map((snapshot) => snapshot.data),
+        aggregator,
+      );
+
+      this.logger.trace('aggregatedResult', aggregatedResult);
+
+      await aggregatedResult.map((result) => {
+        const snapshotHash = latestSnapshot
+          .map((snapshot) => {
+            return createHash('sha256')
+              .update(JSON.stringify(snapshot.data))
+              .digest('hex');
+          })
+          .getOrElse('');
+
+        const toCacheHash = createHash('sha256')
+          .update(JSON.stringify(result))
+          .digest('hex');
+        if (snapshotHash !== toCacheHash) {
+          this.logger.trace('save to cache', result);
+          return this.cache.set(id, { data: result, time: aggregatedAt.toISOString() });
+        }
+      });
+
+      this.logger.trace('aggregateLatency', {
+        query,
+        args,
+        query_time: aggregatedAt.getTime() - start,
+        aggregate_time: Date.now() - aggregatedAt.getTime(),
+        total_time: Date.now() - start,
+      });
+
+      return aggregatedResult;
+    };
+
+    return _impl;
+  }
+
+  public async listen(event_namespace: string, event_type: string, handler: EventHandler<Q, any>): Promise<void> {
+    const pattern = [event_namespace, event_type].join('.');
+
+    const _handler = async (event: Event<any, any>) => {
```

Comment:
Same as in the previous point, because the event generics are bounded to explicit implementations of `EventData` and `EventContext` the any here allows any implementation of those and doesn`t really hinder the type safety.